src/java.base/share/native/libjava/io_util.c

```diff
@@ -142,63 +146,60 @@ writeSingle(JNIEnv *env, jobject this, jint byte, jboolean append, jfieldID fid)
     }
 }
 
-void
+int
 writeBytes(JNIEnv *env, jobject this, jbyteArray bytes,
            jint off, jint len, jboolean append, jfieldID fid)
 {
-    jint n;
+    jint nwritten = -1;
     char stackBuf[BUF_SIZE];
-    char *buf = NULL;
+    char *buf = stackBuf;
     FD fd;
 
     if (IS_NULL(bytes)) {
         JNU_ThrowNullPointerException(env, NULL);
-        return;
+        return -1;
     }
 
     if (outOfBounds(env, off, len, bytes)) {
         JNU_ThrowByName(env, "java/lang/IndexOutOfBoundsException", NULL);
-        return;
+        return -1;
     }
 
     if (len == 0) {
-        return;
+        return 0;
     } else if (len > BUF_SIZE) {
-        buf = malloc(len);
+        if (len > MAX_MALLOC_SIZE)
+            len = MAX_MALLOC_SIZE;
+        buf = (char*)malloc(len*sizeof(char));
         if (buf == NULL) {
             JNU_ThrowOutOfMemoryError(env, NULL);
-            return;
+            return 0;
         }
-    } else {
-        buf = stackBuf;
     }
 
     (*env)->GetByteArrayRegion(env, bytes, off, len, (jbyte *)buf);
 
     if (!(*env)->ExceptionOccurred(env)) {
-        off = 0;
-        while (len > 0) {
-            fd = getFD(env, this, fid);
-            if (fd == -1) {
-                JNU_ThrowIOException(env, "Stream Closed");
-                break;
-            }
-            if (append == JNI_TRUE) {
-                n = IO_Append(fd, buf+off, len);
-            } else {
-                n = IO_Write(fd, buf+off, len);
-            }
-            if (n == -1) {
-                JNU_ThrowIOExceptionWithLastError(env, "Write error");
-                break;
-            }
-            off += n;
-            len -= n;
+        fd = getFD(env, this, fid);
+        if (fd == -1) {
+            JNU_ThrowIOException(env, "Stream Closed");
+            nwritten = -1;
+        } else {
+             if (append == JNI_TRUE) {
+                 nwritten = IO_Append(fd, buf, len);
+             } else {
+                 nwritten = IO_Write(fd, buf, len);
+             }
+             if (nwritten == -1) {
+                 JNU_ThrowIOExceptionWithLastError(env, "Write error");
+             }
         }
     }
-    if (buf != stackBuf) {
+
+    if (buf != stackBuf)
```

Comment:
Wouldn't this cause a leak when if-condition is not met and `free(buf)` is not called?

Reply:
Then I guess we don't need this `if`-clause

Reply:
Then I think one gets an error if `0 < len < BUF_SIZE`:

```
$ cat free.c
#include <stdlib.h>

int main(int argc, char** argv)
{
    char stackBuf[8];
    char* buf;

    buf = stackBuf;
    free(buf);
}
bpb:test{175}$ ./free
free(14365,0x10e6a1600) malloc: *** error for object 0x7ff7b7d38280: pointer being freed was not allocated
free(14365,0x10e6a1600) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```
This is on macOS.

Reply:
Right!